handle consts with param/infer in `const_eval_resolve` better

compiler-errors opened the pull request:
This PR addresses [this thread here](https://github.com/rust-lang/rust/pull/99449#discussion_r924141230). Was this the change you were looking for @lcnr?

Interestingly, one test has begun to pass. Was that expected?

r? @lcnr

compiler-errors committed:
handle consts with param/infer in const_eval_resolve better

compiler-errors commented on compiler/rustc_infer/src/infer/mod.rs:
```diff
@@ -1694,10 +1694,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+            match ac {
+                Ok(None) => {
+                    substs = InternalSubsts::identity_for_item(self.tcx, unevaluated.def.did);
+                    param_env = self.tcx.param_env(unevaluated.def.did);
```
i think lcnr mentioned that the param env should be set since we're using the identity substs, so doing that here. i dont think this affects behavior currently, but it could in the future.

lcnr commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -185,15 +185,7 @@ pub fn is_const_evaluatable<'cx, 'tcx>(
-                    .delay_span_bug(span, &format!("unexpected `TooGeneric` for {:?}", uv));
-                NotConstEvaluatable::MentionsParam
-            } else {
+            Err(ErrorHandled::TooGeneric) => Err({
```
can you move the `Err` into the block or directly use the `guar` in `NotConstEvaluatable::Error` without first assigning it to a local variable?

lcnr commented on compiler/rustc_infer/src/infer/mod.rs:
```diff
@@ -1694,10 +1694,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                                // actually show up in the abstract const's leafs.
+                                self.tcx
+                                    .mk_ty(ty::Placeholder(ty::PlaceholderType {
+                                        universe: ty::UniverseIndex::MAX,
```
you should check here that the generic argument kind is correct, e.g. placeholder type for types and and a placeholder const for constants.

Also, while it doesn't matter, please use the root universe (or the highest universe of all inference variables in that argument) instead of the max universe, as that's closer to what we actually mean.

lcnr commented:
2 nits, then r=me

thanks :heart:

compiler-errors committed:
Address nits, move substs replacement to separate function

compiler-errors commented:
@lcnr can you take one last look at this? 

addressed the nits in the last commit, but I didn't know what to do when we had a `ty::GenericArgKind::Const` whose `ty` references params or infer -- do we want to replace that with a placeholder too? that's [what I did](https://github.com/rust-lang/rust/pull/99618/commits/66d85a9df3f78a0a0f1b9ba62b8df475f2480a9a#diff-79cce3ca916fa3f7d4f1b1879dbec1b3852414a5cfc719087851635aaa240a97R2052), but idk if it's necessarily correct. :thinking:

lcnr commented on compiler/rustc_infer/src/infer/mod.rs:
```diff
@@ -2052,3 +2026,46 @@ impl<'tcx> fmt::Debug for RegionObligation<'tcx> {
+                if ct.has_infer_types_or_consts() || ct.has_param_types_or_consts() =>
+            {
+                let mut ty = ct.ty();
+                // If the type references param or infer, replace that too...
```
that should be unreachable, so you should be able to `bug!` there

compiler-errors replied:
mkay

compiler-errors commented:
@bors r=lcnr

bors commented:
:pushpin: Commit b248647ef0aed88dc8f3be29bec0b98d8be64c3c has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr b248647ef0aed88dc8f3be29bec0b98d8be64c3c -->
<!-- homu: {"type":"Approved","sha":"b248647ef0aed88dc8f3be29bec0b98d8be64c3c","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
